Merge HeliReturnToBase into ReturnToBase

reaperrr opened the pull request:
Most other aircraft activities will have to be merged in a single PR and/or after the idle aircraft refactor, since they're too intertwined to feasibly split them across multiple PRs.
But the RTB activities were comparatively self-contained, so I decided to pull this in before the other aircraft refactorings to make the later PRs a bit smaller.

Depends on #16206.
Sort of split from #15803 (with some improvements, and some changes due to other activities not being merged yet).

Fixes #16377.

reaperrr committed:
Merge HeliReturnToBase into ReturnToBase

reaperrr committed:
Exclude dead actors from ChooseResupplier

pchote commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -135,10 +140,47 @@ public override Activity Tick(Actor self)
+							var randomPosition = WVec.FromPDF(self.World.SharedRandom, 2) * distanceLength / 1024;
+							var target = Target.FromPos(nearestResupplier.CenterPosition + randomPosition);
+
+							Queue(new Fly(self, target, WDist.Zero, aircraft.Info.WaitDistanceFromResupplyBase));
```
Are you sure about using `Queue` here? I haven't looked at the code changes in detail, but I expect this (as opposed to `QueueChild`) would put these after any other player-queued activities, which is probably not what we want.

reaperrr replied:
Judging by #16246, you're right. I also think it would be better to wait for #16246 to be merged first so I only have to update once, though.

reaperrr committed:
Improve ReturnToBase Activity.cs adherance

And make it use child activities for queueability.

reaperrr commented:
Rebased & updated.

@tovl Could you help reviewing this, especially the queueing part? I'm not 100% confident in that myself.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -135,10 +146,40 @@ public override Activity Tick(Actor self)
 				var nearestResupplier = ChooseResupplier(self, false);
```
Line 144: Doesn't `ChooseResupplier` already guarantee that `dest.IsDead` is false?

reaperrr replied:
Yeah, updated.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -135,10 +146,40 @@ public override Activity Tick(Actor self)
+							var randomPosition = WVec.FromPDF(self.World.SharedRandom, 2) * distanceLength / 1024;
+							var target = Target.FromPos(nearestResupplier.CenterPosition + randomPosition);
+
+							QueueChild(self, new HeliFly(self, target, WDist.Zero, aircraft.Info.WaitDistanceFromResupplyBase, targetLineColor: Color.Green));
```
This should probably use `pretick == true` to prevent a one-tick stutter. Same goes for the other instances of `QueueChild` here.

reaperrr replied:
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -135,10 +146,40 @@ public override Activity Tick(Actor self)
+				}
+				else if (nearestResupplier == null && aircraft.Info.VTOL && aircraft.Info.LandWhenIdle)
+				{
+					// Using Queue instead of QueueChild here is intentional, as we want VTOLs with LandWhenIdle to land and stay there in this situation
```
It might be nicer to have them return to a fallback position (such as the player spawn point or construction yard) instead of landing on the spot as that is usually the exact opposite of what you want when ordering return to base in such a situation. This is probably slightly out of scope though.

reaperrr replied:
I agree about the idea, but that's what #15873 (or a follow-up to that) will take care of, so I'd like to leave this as-is in this PR.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -147,28 +188,47 @@ public override Activity Tick(Actor self)
 
-			// Fix a problem when the airplane is send to resupply near the airport
-			landingProcedures.Add(new Fly(self, Target.FromPos(w3), WDist.Zero, new WDist(turnRadius / 2)));
+				// Fix a problem when the airplane is sent to resupply near the airport
```
Line 200 - 206: In the case when `VTOL == true && CanHover == false`, e.g. Harriers, the plane should just approach the airfield in a straight line instead of circling for approach.

reaperrr replied:
Added an `else if` for that case.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -147,28 +188,47 @@ public override Activity Tick(Actor self)
 				landingProcedures.Add(NextActivity);
 
-			return ActivityUtils.SequenceActivities(self, landingProcedures.ToArray());
+			ChildActivity = ActivityUtils.SequenceActivities(self, landingProcedures.ToArray());
```
It's probably nicer to use QueueChild here instead of directly assigning to ChildActivity.

reaperrr replied:
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -147,28 +188,47 @@ public override Activity Tick(Actor self)
+			var exit = dest.FirstExitOrDefault(null);
+			var offset = exit != null ? exit.Info.SpawnOffset : WVec.Zero;
+
 			List<Activity> landingProcedures = new List<Activity>();
```
Do we actually need to make a list of activities here? I think the activities below can just be added to the childactivity queue instead of making a custom list.

reaperrr replied:
> Do we actually need to make a list of activities here?

Not really, that's just a leftover from way back (when this was used to get rid of somewhat messy and less readable SequenceActivities usage).
Fixed.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -814,12 +811,12 @@ public void ResolveOrder(Actor self, Order order)
 				if (!order.Queued)
+				{
 					UnReserve();
+					self.CancelActivity();
```
Why not put `order.Queued` as an argument in Line 819 like you did in Line 798?

reaperrr replied:
I just overlooked that, I think. Updated.

reaperrr commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -127,6 +129,15 @@ public override Activity Tick(Actor self)
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				return this;
+			}
+
```
Done.

reaperrr commented:
Rebased & updated.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -814,12 +811,12 @@ public void ResolveOrder(Actor self, Order order)
 				if (!order.Queued)
+				{
 					UnReserve();
+					self.CancelActivity();
```
This one can be removed now that order.Queued is already forwarded to QueueActivity.

tovl reviewed (commented):
Looks good. I just spotted a few more small things.

reaperrr commented:
Updated.

abcdefg30 reviewed (dismissed):
You can no longer abort a resupply.

tovl commented:
I suspect that the problem is with the fact that `ResupplyAircraft` requeues itself after being cancelled without a queued activity. Because `ReturnToBase` queues it as a childactivity, at the end of the queue, this means `ReturnTobase` keeps going because it's childactivity is never `null`.

Curiously, Line 803-807 in Aircraft.cs already seems to take care of rerunning `ResupplyAircraft` in this case. Yet another instance of two bits of code trying to do the same thing. I think it would be best to remove the offending code from `ResupplyAircraft` and not let that activity bother with this eventuality at all.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -147,28 +191,45 @@ public override Activity Tick(Actor self)
 
 			if (!abortOnResupply)
-				landingProcedures.Add(NextActivity);
+				QueueChild(self, NextActivity, true);
```
Hmm, I don't think it is a good idea to queue `NextActivity` as a childactivity. There should probably just be a `return NextActivity` here somewhere.

reaperrr replied:
I moved the abortOnResupply consideration as well as the `return NextActivity` into the `if (ChildActivity != null)`, more precisely both of these are now triggered when the ChildActivity *becomes* `null`.

reaperrr commented:
Updated.

> I think it would be best to remove the offending code from ResupplyAircraft and not let that activity bother with this eventuality at all.

I think that work-around was done to effectively prevent players from accidentally canceling `ResupplyAircraft` by pressing 'Stop' while the actor resupplies.

I have an idea for a better, unhacky fix for that, but that's out of scope for this PR so I just removed the work-around for now.

matjaeck commented:
Do you want to reopen #15055 for the meantime?

reaperrr commented:
Split the `ResupplyAircraft` changes to #16321, to make it easier to test and review them independently.

reaperrr commented:
Rebased.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -121,6 +122,24 @@ bool ShouldLandAtBuilding(Actor self, Actor dest)
+				// restartAfterChildsFinished serves the purpose of avoiding queueing RTB as ChildActivity of itself
+				// in cases where we want RTB to trigger again after having finished returning to a resupplier
+				// that was reserved when the return was originally triggered.
+				if (ChildActivity != null || restartAfterChildsFinished)
```
I'm confused. How does RTB queue itself as a `ChildActivity`?

reaperrr replied:
Ok, that comment is a bit hard to read (I just wasn't sure how to expess it without making the comment even longer).

What I mean is: Previously, for example https://github.com/OpenRA/OpenRA/pull/16241/files#diff-1f1f9ac8abeec88534accb714c8fbb7bL141 would queue `this` after some other activities, to make the aircraft return to base first and *then* check for any free resuppliers (if there weren't any before).
But queueing `this` feels risky, I don't want to run into an infinite loop. So instead, I use this boolean to basically 'restart' `ReturnToBase` "from the beginning" after the child activities have finished, provided `dest` was null last time we checked and we're still looking for an available resupplier.

pchote replied:
Queuing `this` as a child activity of itself is indeed bogus.

I still don't understand what's going on here.  Is the goal to skip the `abortOnResupply` cancellation?
Why can't we just reset `dest` to null if we want to force a new search?

IMO the first step should be to rewrite this first block as something like
```csharp
if (ChildActivity != null)
{
	ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
	if (ChildActivity != null)
		return this;
}

// resupplied is a new flag set when ResupplyAircraft is queued
if (resupplied && abortOnResupply)
    Cancel(self);
```
so that the child handling follows the same pattern as other activitys, allowing it to be deduplicated.  The next steps after that might become clearer then.

tovl replied:
As far as I can tell, what you're trying to do is already the default. You don't need anything more than the standard `if (ChildActivity != null)` block. Activities queueing `this` after some other activities is (was) by no means an uncommon pattern and it is exactly this pattern that the `ChildActivity` based approach is supposed to replace. Have a look at the code changes in #16348. Is the situation here much different?

reaperrr replied:
@pchote No, the goal is simply to make the activity 'reset' after the child activities finish **if** dest is/was a *reserved* resupplier (because there weren't any unreserved ones when we last checked), and make the activity end if `dest` was not reserved (so we should be done resupplying when the child activities finish).

@tovl The problem here is that with the standard `if (ChildActivity != null)` block without anything else, we'd be lacking a `return NextActivity` for the case of "last search for a resupplier succeeded and we're done resupplying now, so this activity should end now", as the only other `return NextActivity` (currently) triggers on cancellation only.

reaperrr replied:
But yes, @pchote's example already does the trick. I just had to add `resupplied` to one of the `if (...) -> return NextActivity` blocks.

tovl replied:
Ah, now I get it. What you have right now looks like it should do the trick.

reaperrr commented:
Rebased.

reaperrr commented:
Updated.

obrakmann commented:
At the moment I still see at least the following issues:
* Ordering another command while an aircraft (plane or helicopter) is on its way back to base will cause it to go idle
* The logic that makes aircraft queue themselves up for resupply does not work for planes, only for helicopters. It looks like planes never try to get a hold of a reservation when waiting.

And after looking at that Tick method I wonder if it couldn't be split up into three or even four distinct methods, if not even separate activities.

reaperrr commented:
> Ordering another command while an aircraft (plane or helicopter) is on its way back to base will cause it to go idle

What exact effect does giving another order have on ChildActivity? Shouldn't this only happen if the order triggers a `CancelActivity()`?

> The logic that makes aircraft queue themselves up for resupply does not work for planes, only for helicopters. It looks like planes never try to get a hold of a reservation when waiting.

Hm, is this a regression vs. bleed?

obrakmann commented:
> What exact effect does giving another order have on ChildActivity? Shouldn't this only happen if the order triggers a CancelActivity()?

Yes, but it would do that unless you hold shift.

> Hm, is this a regression vs. bleed?

Yes, unfortunately.

reaperrr commented:
> And after looking at that Tick method I wonder if it couldn't be split up into three or even four distinct methods, if not even separate activities.

Definitely, but I think most (larger) aircraft activities could use an overhaul to a varying degree, it's just that I'd like to get the merging out of the way first, so we can get rid of all these damn `if CanHover/VTOL -> queue HeliActivity, else queue [Plane]Activity` places in `Aircraft` and shift most (ideally all) of these CanHover/VTOL considerations to the activities themselves.

reaperrr commented:
> Yes, unfortunately.

This is actually `FlyCircle`'s fault. It only ends after `remainingTicks` reaches zero if there's a `NextActivity`, but due to the switch to child activities on RTB, there (usually) isn't, so `FlyCircle` loops on forever.
Removing that `NextActivity != null` check from FlyCircle's first 'if' block should be fine, since we do have fallbacks for the case when an aircraft really goes idle.

pchote commented:
Would queuing a 0-tick `Wait` work as a hack?

reaperrr commented:
The only place that even uses that time-limit at all is RTB anyway, so just removing that `NextActivity != null` looks like the correct fix to me.

tovl commented:
It should still check for `NextActivity` if ticks is set to -1 though.

reaperrr commented:
Why? It doesn't on bleed, and I don't see were we would need that.

reaperrr commented:
Updated with what I have, also added a commit that makes the current child activity visible with ActorTags checked.

reaperrr commented:
> It should still check for NextActivity if ticks is set to -1 though.

You were right, of course, this is what caused the issue @matjaeck pointed out.

Updated, should be fixed without regressions (fingers crossed...).

reaperrr committed:
Make aircraft without TakeOffOnResupply always land

on resupplier, even if ammo is full, when given a
"ReturnToBase" order via deploy key.
In other words, in that case treat the RTB order like an
explicit "Repair" or "Enter" order.

reaperrr commented:
Minor addition: Aircraft with `TakeOffOnResupply: false` (for example RA MiG and Yak) *and full ammo* now land on a free resupplier when given "ReturnToBase" order via command bar / hotkey, instead of circling over some free resupplier just because they don't need to reload.
Basically, for aircraft with `TakeOffOnResupply: false`, "ReturnToBase" now behaves more like "Enter".

reaperrr committed:
Fix ReturnToBase restarting on each hotkey press

Now that the RTB process is a single activity with childs,
it's relatively easy to prevent the activity from restarting
every time the deploy/RTB hotkey is pressed.

reaperrr commented:
Updated, added another small fix: the `ReturnToBase` activity is no longer restarted on "ReturnToBase" order/hotkey if the activity is already running, preventing additional key presses from delaying the arrival.

reaperrr committed:
Make actors ignore ReturnToBase order if already on resupplier

reaperrr commented:
Since it's another easy one, I added a fix for #16377 as well. Should be the last scope-creep though, anything else should be deferred to a follwo-up.

tovl reviewed (approved):
Looks good to me.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/233d6b21505240cc4340b7edf952666d97221411...0703da50493f42c921da3f5b74e4be2ba33aa78c
